File: frontend/src/lib/historyGroups.test.ts
import { describe, it, expect } from "vitest";
import {
    groupHistory,
    groupKey,
    groupLabel,
    summarizeGroup,
    type GroupableHistoryItem,
} from "./historyGroups";

const item = (over: Partial<GroupableHistoryItem> & { date: Date }): GroupableHistoryItem => ({
    type: "expense",
    amountHome: 10,
    ...over,
});

describe("groupKey", () => {
    it("buckets by local day, month and year", () => {
        const d = new Date(2026, 2, 9, 23, 30);
        expect(groupKey(d, "day")).toBe("2026-03-09");
        expect(groupKey(d, "month")).toBe("2026-03");
        expect(groupKey(d, "year")).toBe("2026");
    });

    it("keeps a late-evening entry on the day the user filed it", () => {
        // A 23:00 local timestamp is the next UTC day in +02:00; keying off
        // local parts keeps it in the bucket the user sees it under.
        const late = new Date(2026, 0, 31, 23, 0);
        expect(groupKey(late, "day")).toBe("2026-01-31");
    });
});

describe("groupLabel", () => {
    const now = new Date(2026, 7, 20, 9, 0);

    it("says Today and Yesterday for day buckets", () => {
        expect(groupLabel(new Date(2026, 7, 20, 18, 0), "day", now)).toMatch(/^Today · /);
        expect(groupLabel(new Date(2026, 7, 19, 1, 0), "day", now)).toMatch(/^Yesterday · /);
        expect(groupLabel(new Date(2026, 7, 18), "day", now)).not.toMatch(/Today|Yesterday/);
    });

    it("does not use Today/Yesterday for month and year buckets", () => {
        expect(groupLabel(new Date(2026, 7, 20), "month", now)).not.toMatch(/Today|Yesterday/);
        expect(groupLabel(new Date(2026, 7, 20), "year", now)).toBe("2026");
    });
});

describe("summarizeGroup", () => {
    it("splits inflow from outflow and nets them", () => {
        const s = summarizeGroup([
            item({ date: new Date(2026, 7, 20), type: "income", amountHome: 3000 }),
            item({ date: new Date(2026, 7, 20), type: "expense", amountHome: 42.5 }),
            item({ date: new Date(2026, 7, 20), type: "buy", amountHome: 500 }),
            item({ date: new Date(2026, 7, 20), type: "sell", amountHome: 100 }),
        ]);
        expect(s.inflow).toBe(3100);
        expect(s.outflow).toBe(542.5);
        expect(s.net).toBe(2557.5);
    });

    it("ignores transfers on both legs", () => {
        // Moving 1000 between your own accounts is neither income nor spending.
        const s = summarizeGroup([
            item({ date: new Date(2026, 7, 20), type: "transfer_out", amountHome: 1000 }),
            item({ date: new Date(2026, 7, 20), type: "transfer_in", amountHome: 1000 }),
            item({ date: new Date(2026, 7, 20), type: "expense", amountHome: 20 }),
        ]);
        expect(s.inflow).toBe(0);
        expect(s.outflow).toBe(20);
        expect(s.net).toBe(-20);
    });

    it("counts rows it cannot convert instead of summing them at face value", () => {
        const s = summarizeGroup([
            item({ date: new Date(2026, 7, 20), type: "expense", amountHome: null }),
            item({ date: new Date(2026, 7, 20), type: "expense", amountHome: 20 }),
        ]);
        expect(s.outflow).toBe(20);
        expect(s.unconverted).toBe(1);
    });

    it("uses magnitudes, so a negative stored amount still reads as spending", () => {
        const s = summarizeGroup([item({ date: new Date(2026, 7, 20), type: "expense", amountHome: -30 })]);
        expect(s.outflow).toBe(30);
        expect(s.net).toBe(-30);
    });

    it("is all zeroes for a group of nothing but transfers", () => {
        const s = summarizeGroup([item({ date: new Date(2026, 7, 20), type: "transfer_in", amountHome: 500 })]);
        expect(s).toEqual({ inflow: 0, outflow: 0, net: 0, unconverted: 0 });
    });
});

describe("groupHistory", () => {
    const items = [
        item({ date: new Date(2026, 7, 20, 10, 0), type: "income", amountHome: 100 }),
        item({ date: new Date(2026, 7, 20, 8, 0), type: "expense", amountHome: 30 }),
        item({ date: new Date(2026, 7, 19, 8, 0), type: "expense", amountHome: 12 }),
        item({ date: new Date(2026, 6, 2, 8, 0), type: "expense", amountHome: 7 }),
        item({ date: new Date(2025, 6, 2, 8, 0), type: "income", amountHome: 5 }),
    ];

    it("buckets by day with a total per day", () => {
        const groups = groupHistory(items, "day", new Date(2026, 7, 20));
        expect(groups.map(g => g.key)).toEqual(["2026-08-20", "2026-08-19", "2026-07-02", "2025-07-02"]);
        expect(groups[0].items).toHaveLength(2);
        expect(groups[0].summary).toMatchObject({ inflow: 100, outflow: 30, net: 70 });
        expect(groups[1].summary).toMatchObject({ outflow: 12, net: -12 });
    });

    it("rolls the same items up by month", () => {
        const groups = groupHistory(items, "month", new Date(2026, 7, 20));
        expect(groups.map(g => g.key)).toEqual(["2026-08", "2026-07", "2025-07"]);
        expect(groups[0].summary).toMatchObject({ inflow: 100, outflow: 42, net: 58 });
    });

    it("rolls the same items up by year", () => {
        const groups = groupHistory(items, "year", new Date(2026, 7, 20));
        expect(groups.map(g => g.key)).toEqual(["2026", "2025"]);
        expect(groups[0].summary).toMatchObject({ inflow: 100, outflow: 49, net: 51 });
        expect(groups[1].summary).toMatchObject({ inflow: 5, outflow: 0, net: 5 });
    });

    it("preserves the incoming order of groups and of items inside them", () => {
        const groups = groupHistory(items, "day", new Date(2026, 7, 20));
        expect(groups[0].items[0].amountHome).toBe(100);
        expect(groups[0].items[1].amountHome).toBe(30);
    });

    it("returns nothing for an empty list", () => {
        expect(groupHistory([], "month")).toEqual([]);
    });
});

File: android/app/src/test/java/com/ivanlee/financetracker/HistoryGroupsTest.kt
package com.ivanlee.financetracker

import com.ivanlee.financetracker.logic.HistoryEntry
import com.ivanlee.financetracker.logic.HistoryGranularity
import com.ivanlee.financetracker.logic.groupHistory
import com.ivanlee.financetracker.logic.historyGroupLabel
import com.ivanlee.financetracker.logic.homeValue
import com.ivanlee.financetracker.logic.summarizeHistory
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneOffset

/**
 * Grouping + per-section totals for the Activity list (`logic/HistoryGroups.kt`).
 * Twin of the web `frontend/src/lib/historyGroups.test.ts` and iOS's `HistoryGroupsTests.swift` —
 * the three must agree about the same numbers. Every instant is explicit UTC, since that is what
 * the bucketing means.
 */
class HistoryGroupsTest {

    private fun at(year: Int, month: Int, day: Int, hour: Int = 12): Instant =
        LocalDateTime.of(year, month, day, hour, 0).toInstant(ZoneOffset.UTC)

    private fun entry(
        date: Instant,
        isTransfer: Boolean = false,
        isInflow: Boolean = false,
        homeAmount: Double? = 10.0,
    ) = HistoryEntry(date, isTransfer, isInflow, homeAmount)

    // ---- homeValue ----------------------------------------------------------------------

    @Test
    fun `homeValue prefers the stored conversion`() {
        assertEquals(42.0, homeValue(42.0, 60.0, "USD", "SGD")!!, 0.0001)
    }

    @Test
    fun `homeValue falls back to the amount when already in base currency`() {
        assertEquals(60.0, homeValue(null, 60.0, "SGD", "SGD")!!, 0.0001)
    }

    @Test
    fun `homeValue is null for an unconverted foreign amount`() {
        assertNull(homeValue(null, 60.0, "USD", "SGD"))
        assertNull(homeValue(null, 60.0, null, "SGD"))
    }

    @Test
    fun `homeValue is always a magnitude`() {
        assertEquals(42.0, homeValue(-42.0, 0.0, null, "SGD")!!, 0.0001)
    }

    // ---- summarizeHistory ---------------------------------------------------------------

    @Test
    fun `summary splits inflow from outflow and nets them`() {
        val day = at(2026, 8, 20)
        val summary = summarizeHistory(
            listOf(
                entry(day, isInflow = true, homeAmount = 3000.0),
                entry(day, homeAmount = 42.5),
                entry(day, homeAmount = 500.0),
            ),
        )
        assertEquals(3000.0, summary.inflow, 0.0001)
        assertEquals(542.5, summary.outflow, 0.0001)
        assertEquals(2457.5, summary.net, 0.0001)
        assertTrue(summary.showsNet)
    }

    @Test
    fun `summary ignores both legs of a transfer`() {
        // Moving 1000 between your own accounts is neither income nor spending.
        val day = at(2026, 8, 20)
        val summary = summarizeHistory(
            listOf(
                entry(day, isTransfer = true, isInflow = true, homeAmount = 1000.0),
                entry(day, isTransfer = true, homeAmount = 1000.0),
                entry(day, homeAmount = 20.0),
            ),
        )
        assertEquals(0.0, summary.inflow, 0.0001)
        assertEquals(20.0, summary.outflow, 0.0001)
        assertEquals(-20.0, summary.net, 0.0001)
    }

    @Test
    fun `summary counts rows it cannot convert instead of summing them`() {
        val day = at(2026, 8, 20)
        val summary = summarizeHistory(listOf(entry(day, homeAmount = null), entry(day, homeAmount = 20.0)))
        assertEquals(20.0, summary.outflow, 0.0001)
        assertEquals(1, summary.unconverted)
    }

    @Test
    fun `summary hides the net when only one side moved`() {
        val day = at(2026, 8, 20)
        assertFalse(summarizeHistory(listOf(entry(day, homeAmount = 20.0))).showsNet)
        assertFalse(summarizeHistory(listOf(entry(day, isInflow = true, homeAmount = 20.0))).showsNet)
    }

    @Test
    fun `summary of nothing but transfers is all zeroes`() {
        val summary = summarizeHistory(listOf(entry(at(2026, 8, 20), isTransfer = true, homeAmount = 500.0)))
        assertEquals(0.0, summary.inflow, 0.0001)
        assertEquals(0.0, summary.outflow, 0.0001)
        assertEquals(0, summary.unconverted)
    }

    // ---- groupHistory -------------------------------------------------------------------

    private val sample
        get() = listOf(
            entry(at(2026, 8, 20, 10), isInflow = true, homeAmount = 100.0),
            entry(at(2026, 8, 20, 8), homeAmount = 30.0),
            entry(at(2026, 8, 19, 8), homeAmount = 12.0),
            entry(at(2026, 7, 2, 8), homeAmount = 7.0),
            entry(at(2025, 7, 2, 8), isInflow = true, homeAmount = 5.0),
        )

    @Test
    fun `groups by day with a total per day`() {
        val groups = groupHistory(sample, HistoryGranularity.DAY) { it }
        assertEquals(4, groups.size)
        assertEquals(2, groups[0].items.size)
        assertEquals(at(2026, 8, 20, 0), groups[0].start)
        assertEquals(100.0, groups[0].summary.inflow, 0.0001)
        assertEquals(30.0, groups[0].summary.outflow, 0.0001)
        assertEquals(12.0, groups[1].summary.outflow, 0.0001)
    }

    @Test
    fun `rolls the same items up by month`() {
        val groups = groupHistory(sample, HistoryGranularity.MONTH) { it }
        assertEquals(3, groups.size)
        assertEquals(at(2026, 8, 1, 0), groups[0].start)
        assertEquals(100.0, groups[0].summary.inflow, 0.0001)
        assertEquals(42.0, groups[0].summary.outflow, 0.0001)
        assertEquals(58.0, groups[0].summary.net, 0.0001)
    }

    @Test
    fun `rolls the same items up by year`() {
        val groups = groupHistory(sample, HistoryGranularity.YEAR) { it }
        assertEquals(2, groups.size)
        assertEquals(at(2026, 1, 1, 0), groups[0].start)
        assertEquals(49.0, groups[0].summary.outflow, 0.0001)
        assertEquals(51.0, groups[0].summary.net, 0.0001)
        assertEquals(5.0, groups[1].summary.inflow, 0.0001)
    }

    @Test
    fun `preserves the incoming order of groups and items`() {
        val groups = groupHistory(sample, HistoryGranularity.DAY) { it }
        assertEquals(groups.map { it.start }.sortedDescending(), groups.map { it.start })
        assertEquals(100.0, groups[0].items[0].homeAmount!!, 0.0001)
        assertEquals(30.0, groups[0].items[1].homeAmount!!, 0.0001)
    }

    @Test
    fun `groups nothing into nothing`() {
        assertTrue(groupHistory(emptyList<HistoryEntry>(), HistoryGranularity.MONTH) { it }.isEmpty())
    }

    // ---- labels -------------------------------------------------------------------------

    @Test
    fun `day labels say Today and Yesterday relative to the given now`() {
        val now = at(2026, 8, 20, 9)
        assertTrue(historyGroupLabel(at(2026, 8, 20, 0), HistoryGranularity.DAY, now).startsWith("Today · "))
        assertTrue(historyGroupLabel(at(2026, 8, 19, 0), HistoryGranularity.DAY, now).startsWith("Yesterday · "))
        val older = historyGroupLabel(at(2026, 8, 18, 0), HistoryGranularity.DAY, now)
        assertFalse(older.contains("Today"))
        assertFalse(older.contains("Yesterday"))
    }

    @Test
    fun `month and year labels never say Today or Yesterday`() {
        val now = at(2026, 8, 20, 9)
        val month = historyGroupLabel(at(2026, 8, 1, 0), HistoryGranularity.MONTH, now)
        assertFalse(month.contains("Today"))
        assertTrue(month.contains("2026"))
        assertEquals("2026", historyGroupLabel(at(2026, 1, 1, 0), HistoryGranularity.YEAR, now))
    }

    @Test
    fun `a late-evening UTC entry stays on its own day`() {
        // 23:00 UTC is the next day in +02:00; bucketing in UTC keeps it where the backend put it.
        val groups = groupHistory(listOf(entry(at(2026, 1, 31, 23))), HistoryGranularity.DAY) { it }
        assertEquals(at(2026, 1, 31, 0), groups[0].start)
    }
}
